Teach pytrees about namedtuple

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#62292 Teach pytrees about namedtuple**

This PR adds pytree support for namedtuples. The challenge about namedtuple
is that each namedtuple class is actually different. This PR does the
following:
- it adds a namedtuple flatten/unflatten. The flatten function returns
a context that is the actual type of the namedtuple subclass. The
unflatten function uses that type to reconstruct the namedtuple
- Special cases all pytree logic to consider all namedtuples the same.
This is done by creating a `_get_node_type(pytree)` helper function that
returns `namedtuple` if `pytree` is any namedtuple subclass. The effect
of this is that all namedtuple subclasses will go through the namedtuple
flatten/unflatten functions
- Adds a `_namedtuple_flatten_spec` function for FX pytrees. This function
flattens the namedtuple based on the spec and is equivalent to the
`_tuple_flatten_spec`.

Test Plan
- new tests in test/test_pytree.py and test/test_fx.py

Differential Revision: [D29947302](https://our.internmc.facebook.com/intern/diff/D29947302)

zou3519 committed:
Teach pytrees about namedtuple

This PR adds pytree support for namedtuples. The challenge about namedtuple
is that each namedtuple class is actually different. This PR does the
following:
- it adds a namedtuple flatten/unflatten. The flatten function returns
a context that is the actual type of the namedtuple subclass. The
unflatten function uses that type to reconstruct the namedtuple
- Special cases all pytree logic to consider all namedtuples the same.
This is done by creating a `_get_node_type(pytree)` helper function that
returns `namedtuple` if `pytree` is any namedtuple subclass. The effect
of this is that all namedtuple subclasses will go through the namedtuple
flatten/unflatten functions
- Adds a `_namedtuple_flatten_spec` function for FX pytrees. This function
flattens the namedtuple based on the spec and is equivalent to the
`_tuple_flatten_spec`.

Test Plan
- new tests in test/test_pytree.py and test/test_fx.py

[ghstack-poisoned]

zou3519 committed:
Update on "Teach pytrees about namedtuple"


This PR adds pytree support for namedtuples. The challenge about namedtuple
is that each namedtuple class is actually different. This PR does the
following:
- it adds a namedtuple flatten/unflatten. The flatten function returns
a context that is the actual type of the namedtuple subclass. The
unflatten function uses that type to reconstruct the namedtuple
- Special cases all pytree logic to consider all namedtuples the same.
This is done by creating a `_get_node_type(pytree)` helper function that
returns `namedtuple` if `pytree` is any namedtuple subclass. The effect
of this is that all namedtuple subclasses will go through the namedtuple
flatten/unflatten functions
- Adds a `_namedtuple_flatten_spec` function for FX pytrees. This function
flattens the namedtuple based on the spec and is equivalent to the
`_tuple_flatten_spec`.

Test Plan
- new tests in test/test_pytree.py and test/test_fx.py

[ghstack-poisoned]

zou3519 commented:
@zou3519 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29947302).

zou3519 commented:
Adding reviewers from original pytree PR: https://github.com/pytorch/pytorch/pull/46287

albanD commented on test/test_fx.py:
```diff
@@ -65,6 +66,11 @@ def forward(self, x):
 def a_non_torch_leaf(a, b):
     return a + b
 
+# used in test_pytree. It's all the way out here because picking a GraphModule
```
pickling?

zou3519 replied:
Yes, should be pickling, thanks for the catch

albanD commented on test/test_pytree.py:
```diff
@@ -59,6 +60,33 @@ def run_test(tup):
+            self.assertEqual(treespec, expected_spec)
+
+            unflattened = tree_unflatten(values, treespec)
+            self.assertEqual(unflattened, tup)
```
How deep is assertEqual doing for nested structures?

zou3519 replied:
I can't find the source code responsible for this, but by some testing, it looks like the answer is pretty deep:
```
(Pdb) self.assertEqual((1, 2), (2, 3))
*** AssertionError: False is not true : Scalars failed to compare as equal! 1 != 2
(Pdb) self.assertEqual((torch.tensor(3),), (torch.tensor(4),))
*** AssertionError: False is not true : Tensors failed to compare as equal!Found 1 different element(s)
 (out of 1), with the greatest difference of 1 (3 vs. 4) occuring at index 0.
 (Pdb) self.assertEqual(([torch.tensor(3)],), ([torch.tensor(4)],))
*** AssertionError: False is not true : Tensors failed to compare as equal!Found 1 different element(s)
 (out of 1), with the greatest difference of 1 (3 vs. 4) occuring at index 0.
 ```

albanD commented on torch/utils/_pytree.py:
```diff
@@ -56,14 +57,38 @@ def _tuple_flatten(d: Tuple[Any, ...]) -> Tuple[List[Any], Context]:
     return tuple(values)
 
+def _namedtuple_flatten(d: NamedTuple) -> Tuple[List[Any], Context]:
+    return list(tuple(d)), type(d)
```
You can't just call list on a namedtuple? You do that in the tests no?

zou3519 replied:
Good question. Turns out that you can:
```
from collections import namedtuple
Point = namedtuple('Point', ['x', 'y'])
somewhere = Point(x=0, y=1)
print(list(somewhere))
```

albanD reviewed (approved):
LGTM

zou3519 committed:
Update on "Teach pytrees about namedtuple"


This PR adds pytree support for namedtuples. The challenge about namedtuple
is that each namedtuple class is actually different. This PR does the
following:
- it adds a namedtuple flatten/unflatten. The flatten function returns
a context that is the actual type of the namedtuple subclass. The
unflatten function uses that type to reconstruct the namedtuple
- Special cases all pytree logic to consider all namedtuples the same.
This is done by creating a `_get_node_type(pytree)` helper function that
returns `namedtuple` if `pytree` is any namedtuple subclass. The effect
of this is that all namedtuple subclasses will go through the namedtuple
flatten/unflatten functions
- Adds a `_namedtuple_flatten_spec` function for FX pytrees. This function
flattens the namedtuple based on the spec and is equivalent to the
`_tuple_flatten_spec`.

Test Plan
- new tests in test/test_pytree.py and test/test_fx.py

Differential Revision: [D29947302](https://our.internmc.facebook.com/intern/diff/D29947302)

[ghstack-poisoned]

zou3519 commented:
@zou3519 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29947302).
